backend/cfg/cfg.ml

```diff
@@ -32,62 +32,207 @@ include Cfg_intf.S
 module BasicInstructionList = struct
   type instr = basic instruction
 
-  type cell =
+  type node =
     { instr : instr;
-      mutable before_rev : instr list;
-      mutable after : instr list
+      mutable prev : node;
+      mutable next : node
     }
 
-  let insert_before cell instr = cell.before_rev <- instr :: cell.before_rev
-
-  let insert_after cell instr = cell.after <- instr :: cell.after
-
-  let instr cell = cell.instr
-
-  type t = instr list ref
-
-  let make_empty () = ref []
-
-  let make_single instr = ref [instr]
-
-  let of_list l = ref l
-
-  let hd t = match !t with [] -> None | hd :: _ -> Some hd
+  (* CR xclerc for xclerc: a dummy instruction value has probably been
+     introduced by another pull request. *)
+  let dummy_instruction : instr =
+    { desc = Prologue;
+      arg = [||];
+      res = [||];
+      dbg = Debuginfo.none;
+      fdo = Fdo_info.none;
+      live = Reg.Set.empty;
+      stack_offset = -1;
+      id = -1;
+      irc_work_list = Unknown_list
+    }
 
-  let add_begin t instr = t := instr :: !t
+  let rec dummy_node =
+    { instr = dummy_instruction; prev = dummy_node; next = dummy_node }
 
-  let add_end t instr = t := !t @ [instr]
+  let[@inline] unattached_node instr =
+    { instr; prev = dummy_node; next = dummy_node }
 
-  let is_empty t = match !t with [] -> true | _ :: _ -> false
+  type t =
+    { mutable length : int;
+      mutable first : node;
+      mutable last : node
+    }
```

Comment:
Wouldn't it be simpler to have a sentinel node, and the definition of `t` be:
`type t = { mutable length : int; mutable sentinel : node }`?
Then `first` and `last` are accessed using `t.sentinel.next` and `t.sentinel.prev`, and some functions become a lot simpler as well, since we don't need to test for comparison with `dummy_node` to check whether to update `t.first` or `t.last`.

Reply:
Possibly, will have a closer look.

Reply:
@Ekdohibs  What did you have in mind for representing an empty list using a sentinel? I mean, some comparisons may be needed or the invariant between length and the list representation relaxed. Does sentinel need to be mutable?